Some group theory tools for Quantum error correction

IsaacP1234 opened the pull request:
Implemented groupify function to find group of Paulis generated by a generating set.

Implemented get_generating_set function to find minimal generating set of a group of Paulis.

Implemented normalize function to find all Paulis that commute with a given set of paulis.

Implemented center function to find all Paulis within a set that commute with all other elements of that set.

Added united tests for all four functions.

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+# `invert' arbitrary groups
+
+"""
+input: (generating) set of PauliOperator
```
Check other docstrings in the library, e.g. at http://qc.quantumsavory.org/stable/API/ and follow a similar style.

E.g. a very brief one-line header followed by a paragraph of narative text explaining what is happening. The input and output description can be just part of the text, no need for separate bullets if the function is simple.

IsaacP1234 replied:
Almost all the functions in the API have the one-line header and then some code examples. For simpler functions like groupify and pauligroup, should I include the paragraph portion?

Krastanov replied:
yup, `groupify` will not be simple to someone new to the concept

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+"""
+function groupify(s::Stabilizer) # add checks?
+    n = length(s)
+    group = zero(Stabilizer, 2^n, nqubits(s))
```
Use `Tableau` instead of `Stabilizer` and remove the corresponding TODO.

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+        end
+        return gs
+    else
+        return s[1:r, :]
```
which is fine if given what we said elsewhere about dealing only with real phases

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+For a given non-necessarily-minimal generating set represented as a tableau `s`,
+return the minimal generating set.
+"""
+function get_generating_set(s) # TODO potentially rename to `generator_tableau` or `generating_set`
```
let's rename it to `minimal_generating_set`

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+#= function logical_operator_canonicalize(s)
+    # TODO potentially use
+    # julia> MixedDestabilizer(S"XXX") |> logicalxview
+    # julia> MixedDestabilizer(S"XXX") |> logicalzview
+    # or stabilizerview or destabilizerview
+    # @assert length(s) == length(get_generating_set(groupify(s)))
+    mx = MixedDestabilizer(s)
+    return vcat(logicalxview(mx), logicalzview(mx), s), rank(mx) # TODO use Tableau instead of Stabilizer
+end =#
+
+
+##
+# s = S"XZ ZX IY"
+# s = S"ZZ XZ YY"
+# println("")
+# t = logical_operator_canonicalize(s)
+# ##
+
+#= function commutavise(s::Stabilizer) # TODO take a Tableau
+    @assert length(s) == length(groupify(s))
+    embedding = []
+    # currently this works only on generating sets for abelian groups
+    # maybe split in commutavise
+    # news = ... with n+(n-rank) columns
+    # news[:, 1:n] = s
+    # add some value to news[:, n+1:2n-rank]
+    for P in s
+        for Q in s
+            if comm(P, Q) == 0x01
+                # finish
+            end
+        end
+    end
+end =#
+
+# s = S"XZ ZX IY"
+# s = groupify(s)
+# println("")
+# print(get_generating_set(s))
+# ##
```
remove unfinished things from the pull request so that it can be merged

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+"""
+function pauligroup(n, phases=false)
+    if phases
+        s = zero(Stabilizer, 4^(n + 1), n)
```
let's use a `Tableau` here

Any Tableau can be converted to a Stabilizer by calling `Stabilizer(the_tableau)`

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+"""
+Returns the full pauli group of the given length. Phases besides + are ignored by default, but can be included by setting phases = true. 
+"""
+function pauligroup(n, phases=false)
```
```suggestion
function pauligroup(n; phases=false)
```

That way it is a keyword argument instead of a default argument (makes things a bit more legible)

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+"""
+For a given s, returns the subset of elements in s that commute with all elements in s.
+"""
+function center(s::Stabilizer)  # hilariously inefficient. centerify? centrify?
```
how about naming this to `centralizer`?

Krastanov replied:
Should this take `Stabilizer` because it is meant to be about a state, or `Tableau` because it is for any set of Pauli operators or about an arbitrary Pauli subgroup?

IsaacP1234 replied:
Because stabilizer groups are commutative, running this function on a stabilizer would return the input, so I think it should take in a Tableau. As for the output, the output will be commutative, so I assume returning it as a Stabilizer type is correct.

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+"""
+For a given s, find all Pauli elements of the number of qubits that commute with all elements in s
+"""
+function normalize(s::Stabilizer)
```
How about renaming it to `normalizer`?

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+# print(get_generating_set(s))
+# ##
+"""
+Returns the full pauli group of the given length. Phases besides + are ignored by default, but can be included by setting phases = true. 
```
```suggestion
Returns the full Pauli group of the given length. Phases besides `+` are ignored by default, but can be included by setting `phases = true`. 
```

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+end
+
+"""
+For a given s, find all Pauli elements of the number of qubits that commute with all elements in s
```
not very clear what this says, could you clarify the docstring?

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+end
+
+"""
+For a given s, returns the subset of elements in s that commute with all elements in s.
```
```suggestion
For a given `s`, return the subset of elements in `s` that commute with all elements in `s`.
```

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+    return stabilizer
+end
+
+# s = S"XXII IIXX ZIZI IZIZ"
+# s = groupify(s)
+
+#s = S"II YY"
+#s = groupify(s)
+
+#println(normalize(s))
+
+##
+# Pgroup = pauligroup(2)
+# println(Pgroup)
+##
+
+# scenter = center(s)
+# print(scenter)
+# println(canonicalize!(s)[1:,])
+
+# function center(s::Stabilizer) # centerify? centrify?
+#     #  # returns pairs of non-commuting generators
+# end
+
+# function
```
clean up

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+
+
+
+function contract(s::PauliOperator, subset)
```
probably needs a better name and a docstring

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+    return Stabilizer([P for P in contract.(𝒮, (subset,)) if !isnothing(P)])  # make more efficient
+end =#
+
+function delete(s::PauliOperator, subset)
```
Should be the standard function `Base.deleteat!`, not a new function `delete`. If it is being implemented, it should probably go in the file where `setindex` and `getindex` are defined.

Krastanov replied:
probably in `paulioperators`

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,284 @@
+# s2 = contract(s2, [5])
+
+
+# println(canonicalize!(s))
+# println("______!!!")
+
+# logical operators
+# s = contract(s, [6, 7, 8])
+# s = delete(s, [5])
+#println(canonicalize!(s))
+
+
+# s_contract = contract(s, [1,2])
+
+
+# println("-----")
+
+# s = S"XZ ZX"
+# s = groupify(s)
+
+# s = S"XZZ ZXZ ZZX"
+# s = groupify(s)
+
+# s = Stabilizer(wheel_graph(6))
+# s = groupify(s)
+
+# deleted = delete(s, [1])
+# println(canonicalize!(deleted))
+# println("_---")
+# contracted = contract(s, [1])
+# println(canonicalize!(contracted)[1:4])
+
+# println("---")
+# using QuantumClifford.ECC
+
+# println(canonicalize!(parity_checks(Perfect5())[:,[1, 2, 3, 5, 4]]))
+# println(typeof(groupify(S"XZ ZX")))
+
+
+# # md = MixedDestabilizer(s); groupify(copy(vcat(stabilizerview(md), logicalxview(md), logicalzview(md)))) x
```
clean up

Krastanov commented on CHANGELOG.md:
```diff
@@ -5,6 +5,14 @@
+Implemented get_generating_set function to find minimal generating set of a group of Paulis
+Implemented normalize function to find all Paulis that commute with a given set of paulis
+Implemented center function to find all Paulis within a set that commute with all other elements of that set
+Added united tests for all four functions
```
You can leave CHANGELOG unchanged to avoid merge conflicts until we are done with the reviews. You probably want to rebase this on master:

```
git checkout master
git pull # potentially with specifying the upstream remote
git checkout some_group_functions
git rebase master
```

Krastanov reviewed (changes requested):
Looks like a great start! Let's do some final polish so that we have what is currently available merged, before we worry about more sophisticated functionality. Doing small merges often is much easier for me than doing large pull requests.

Converting to draft for now. Convert back to non-draft and re-request a review when ready.

@KDGoodenough , could you check whether the tests seem reasonable for the functionality you have in mind?

IsaacP1234 commented:
Thank you for the feedback, I think I have implemented it all. 
CHANGELOG left blank to avoid merge conflicts.
Fails JET tests because there are 7 additional warnings, however these are because it interprets indexing a Tableau as a Tableau and not a PauliOperator. I didn't change the limit on warnings in test_jet because that results in a merge conflict. 
I did not rebase this into the master branch of my fork since the pull request seems to work without it and I am not sure what it would help with.

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,239 @@
++ Y
++ Z
++ X
+'''
```
```
See also: [`traceout!`](@ref)
```

Krastanov replied:
add a "see also" pointing to this function in `traceout!`

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,239 @@
+between the given set of qubits and the qubits in the given Stabilizer. Then return 
+the part of the given Stabilizer with the remaining vertical indices.
+
+'''jldoctest
```
fix to backticks

Krastanov reviewed (changes requested):
could you let me know what the new jet issues are (just post some screenshots)

IsaacP1234 commented:
> could you let me know what the new jet issues are (just post some screenshots)

![image](https://github.com/user-attachments/assets/cbe204af-a208-4a04-b794-0deae7a9a7da)

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,239 @@
+    for i in 0:2^n-1
+        for (digit_order, j) in enumerate(digits(i, base=2, pad=n))
+            if j == 1
+                group[i+1] *= s[digit_order]
```
the JET warning stems from the fact that the compiler can not prove that `group[i+1]` and `s[digit_order]` are both `PauliOperator`. The weird thing is that this seems to be due to the fact that it can not prove that `i` is integer, which seems to be not proven because it can not prove that `n` is an integer, because it does not know that the output of `length` is an integer.

The root solution would be to make sure the compiler knows that `length` can only be an integer (wherever it is defined), but just as a workaround for the moment, I am curious, could you just annotate with `n = length(s)::Int` and see whether that particular JET warning disappears.

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,239 @@
+-iYY
+'''
+"""
+function pauligroup(n; phases=false)
```
annotate `n::Int` to see whether the compiler and JET will stop complaining

Krastanov reviewed (changes requested):
Thanks for taking care of this!
I am just responding to the JET reports for the moment. They seem like false positives, but could you please try the following annotations to help the compiler to not raise these false positives. I will do a final review after we have checked that.

When you are done with this, please also make sure to re-request a code review (should be towards the start of the page on the right) to make sure this goes on my automated reading list.

IsaacP1234 commented:
Adding a see also from trace! to delete_columns causes the docs to fail to build, I think because the project_trace_reset.jl docs get created before the group_tableaux.jl ones. 

Additionally, the build fails, but it seems to be an error with the second part, QuantumSavory, and not QuantumClifford. Let me know if this is not the case and I will debug it.

IsaacP1234 commented:
CHANGELOG:
Implemented  `groupify` function to generate stabilizer group from generating set.
Implemented `minimal_generating_set` function to find the minimal generating set of a Stabilizer.
Implemented `pauligroup` function to generate the full Pauli group of a certain length.
Implemented `normalizer` function to generate all Paulis that commute with a set of Paulis
Implemented `centralizer` function to find a subset of a set of Paulis such that each element in the subset commutes with each element in the set. 
Implemented `contractor` function to find a subset of Paulis in a stabilizer that have an identity operator on a certain qubit. 
Implemented `delete_columns` function to remove the operators corresponding to a certain qubit from all Paulis in a Stabilizer.

IsaacP1234 commented:
@Krastanov re-pinging for review as it is now fully complete but a review request is currently active.

Krastanov commented on src/grouptableaux.jl:
```diff
@@ -0,0 +1,246 @@
+            if comm(P, Q) == 0x01
+                commutes = 1
+                break
+            end
+        end
+        if commutes == 0
```
This does not seem to match the description. I would have expected only `+X` to be returned.

Krastanov reviewed (commented):
minor doc cleanup

Krastanov reviewed (changes requested):
I made a few documentation changes. Generally, the documentation should say what something is and why it exists. It should say "how" only if there is something particularly non-trivial and only if the previous two questions are already fully addressed. Frequently that "how" might be just in the code comments. If the "how" is mostly obvious from the code, it should probably not even be a comment. The reasoning behind this is two-fold: (1) the user probably does not care exactly how something is implemented, rather they care what it is; (2) the implementation might change without anyone remembering the update the documentation.

I do have one question for clarification on the contract function. Otherwise the rest looks great.

IsaacP1234 commented:
I have removed the groupify functionality from contractor. PR should be all good.

Krastanov commented:
Thanks, @IsaacP1234 , this was great!

@KDGoodenough , let's see what you are planning to use this for!
